win32ss/user/ntuser/msgqueue.c

```diff
@@ -1871,45 +1872,77 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
         // to switch the keyboard layout and store it to global variable.
         // Using this combination of hotkeys in this function
 
-        if ( gdwLanguageToggleKey < 3 &&
-             IS_KEY_DOWN(gafAsyncKeyState, gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL) )
+        if (gdwLanguageToggleKey < 3)
         {
-            if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
+            LanguageHotkey = gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL;
+            if(IS_KEY_DOWN(gafAsyncKeyState, LanguageHotkey) ||
+               IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT))
             {
-                WPARAM wParamILR;
-                PKL pkl = pti->KeyboardLayout;
+               if (Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
```

Comment:
```suggestion
               if (Msg->wParam == VK_SHIFT && !IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT))
```

Reply:
I would prefer to keep it as is, but if needed i will change it.